includes/Modules/Analytics_4.php

```diff
@@ -826,4 +861,468 @@ public function check_service_entity_access() {
+			$date_range = new Google_Service_AnalyticsData_DateRange();
+			$date_range->setStartDate( $args['start_date'] );
+			$date_range->setEndDate( $args['end_date'] );
+			$request->setDateRanges( array( $date_range ) );
+		}
+
+		$dimension_filter_expressions = array();
+
+		$hostnames = array_values(
+			array_unique(
+				array_map(
+					function ( $site_url ) {
+						return URL::parse( $site_url, PHP_URL_HOST );
+					},
+					$this->permute_site_url( $this->context->get_reference_site_url() )
+				)
+			)
+		);
+
+		$dimension_in_list_filter = new Google_Service_AnalyticsData_InListFilter();
+		$dimension_in_list_filter->setValues( $hostnames );
+		$dimension_filter = new Google_Service_AnalyticsData_Filter();
+		$dimension_filter->setFieldName( 'hostName' );
+		$dimension_filter->setInListFilter( $dimension_in_list_filter );
+		$dimension_filter_expression = new Google_Service_AnalyticsData_FilterExpression();
+		$dimension_filter_expression->setFilter( $dimension_filter );
+		$dimension_filter_expressions[] = $dimension_filter_expression;
+
+		if ( ! empty( $args['dimension_filters'] ) ) {
+			$dimension_filter_expressions = array_merge( $dimension_filter_expressions, $args['dimension_filters'] );
+		}
+
+		if ( ! empty( $args['page'] ) ) {
+			$args['page']            = str_replace( trim( $this->context->get_reference_site_url(), '/' ), '', esc_url_raw( $args['page'] ) );
+			$dimension_string_filter = new Google_Service_AnalyticsData_StringFilter();
+			$dimension_string_filter->setMatchType( 'EXACT' );
+			$dimension_string_filter->setValue( rawurldecode( $args['page'] ) );
+			$dimension_filter = new Google_Service_AnalyticsData_Filter();
+			$dimension_filter->setFieldName( 'pagePathPlusQueryString' );
+			$dimension_filter->setStringFilter( $dimension_string_filter );
+			$dimension_filter_expression = new Google_Service_AnalyticsData_FilterExpression();
+			$dimension_filter_expression->setFilter( $dimension_filter );
+			$dimension_filter_expressions[] = $dimension_filter_expression;
+		}
+
+		$dimension_filter_expression_list = new Google_Service_AnalyticsData_FilterExpressionList();
+		$dimension_filter_expression_list->setExpressions( $dimension_filter_expressions );
+		$dimension_filter_expression = new Google_Service_AnalyticsData_FilterExpression();
+		$dimension_filter_expression->setAndGroup( $dimension_filter_expression_list );
+		$request->setDimensionFilter( $dimension_filter_expression );
+
+		if ( ! empty( $args['row_limit'] ) ) {
+			$request->setLimit( $args['row_limit'] );
+		}
+
+		return $request;
+	}
+
+	/**
+	 * Validates the report metrics.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Google_Service_AnalyticsData_Metric[] $metrics The metrics to validate.
+	 * @throws Invalid_Report_Metrics_Exception Thrown if the metrics are invalid.
+	 */
+	protected function validate_report_metrics( $metrics ) {
+		if ( false === $this->is_using_shared_credentials ) {
```

Comment:
Note: the changes in #5780 are relevant here. We should keep this as-is for now of course because that one is still in definition but I'll make a note on that one that the proposed changes there would apply here as well.